apps/wallet/src/ui/app/components/address-input/validation.ts

```diff
@@ -1,22 +1,42 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-import { isValidSuiAddress } from '@mysten/sui.js';
+import { isSuiNSName, useRpcClient, useSuiNSEnabled } from '@mysten/core';
+import { type JsonRpcProvider, isValidSuiAddress } from '@mysten/sui.js';
+import { useMemo } from 'react';
 import * as Yup from 'yup';
 
-export const SUI_ADDRESS_VALIDATION = Yup.string()
-    .ensure()
-    .trim()
-    .required()
-    .transform((value: string) =>
-        value.startsWith('0x') || value === '' || value === '0'
-            ? value
-            : `0x${value}`
-    )
-    .test(
-        'is-sui-address',
-        // eslint-disable-next-line no-template-curly-in-string
-        'Invalid address. Please check again.',
-        (value) => isValidSuiAddress(value)
-    )
-    .label("Recipient's address");
+export function createSuiAddressValidation(
+    rpc: JsonRpcProvider,
+    suiNSEnabled: boolean
+) {
+    return Yup.string()
+        .ensure()
+        .trim()
+        .required()
+        .test(
+            'is-sui-address',
+            'Invalid address. Please check again.',
+            async (value) => {
+                if (suiNSEnabled && isSuiNSName(value)) {
+                    const address = await rpc.resolveNameServiceAddress({
```

Comment:
This probably can end-up being called multiple times and trigger many requests even for the same value (because other values changed). I guess we are ok for now?

Reply:
Let me add a local cache for this